protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,241 @@
+
+The Tendermint algorithm as described in the [arXiv paper][arxiv],
+and as implemented in CometBFT up to version `v0.38.0-rc3`,
+cannot guarantee consensus _termination_ for applications
+that just fulfill requirement 3b (eventual coherence), but do not fulfill requirements 3, 4, and 5.
+
+We need the following modifications (in terms of the algorithm as described in page 6 of the arXiv paper):
+
+- remove the evaluation of `valid(v)` in lines 29, 36 and 50 (i.e. replace `valid(v)` by `true`)
+- modify line 23 as follows
+
+> _\[Original\]_ &nbsp; 23: **if** $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$ **then**
+
+> _\[Modified\]_
+>
+> &nbsp; 23a: $validValMatch := (validRound_p \neq -1 \land validValue_p = v)$
+>
+> &nbsp; 23b: **if** $[lockedRound_p = −1 \land (validValMatch \lor valid(v))] \lor lockedValue_p=v$ **then**
+
+- The occurrences of `valid(v)` that we have removed were in a way redundant,
+  so removing them does not affect the ability of the algorithm to fulfill consensus properties.
+- Regarding line 23, the changes have the following goals:
+  - if `v` matches the block we have as $validValue_p$ or $lockedValue_p$, we skip the call
+    to `valid(v)`. The reason is that, by the algorithm, if we have `v` as $validValue_p$ or $lockedValue_p$
+    we have received it from at least $2f + 1$ prevotes for $v$ (line 36)
+  - if the previous condition is not met, then we call `valid(v)`
+  - the modification to the conditions must _only_ affect the decision whether to call `valid(v)` or not.
+    It must not modify which part of the "if" should be taken.
+
+Notice we have kept the original `valid(v)` notation, but it stands for the more general $valid(v, bc_{h-1}, x_p)$.
+These algorithmic modifications have also been made to CometBFT (on branch `main`)
+as part of issues [#1171][1171], and [#1230][1230].
+
+## Conclusion
+
+This document has explored the possibility of relaxing the coherence and determinism properties
+of the ABCI 1.0 (and 2.0) specification affecting `PrepareProposal` and `ProcessProposal`
+for a class of applications that cannot guarantee them.
+
+We first weakened the _validity_ property of the consensus specification
+in a way that keeps the overall consensus specification strong enough to be relevant.
+We then proposed a weaker coherence property for ABCI 1.0 (and 2.0) that can replace the original
+coherence and determinism properties related to `PrepareProposal` and `ProcessProposal`.
+The new property is useful for applications that cannot fulfill the original properties
+but can fulfill the new one.
+Finally, we explained how to modify the Tendermint consensus algorithm to guarantee
+the consensus _termination_ property for applications that fulfill the new property.
+
+In this document, we have not tackled the problem of applications
+that cannot fulfill coherence and determinism properties that refer to vote extensions in the ABCI 2.0 specification.
+We leave this as future work.
```

Comment:
I would at least mention that we need to find a way to properly address those changes in the case of a recovering process.

The case to be considered is when the application accepted a value during regular operation and may reject the same value during recovery, or vice-versa. This possibility has the potential of producing conflicting Prevote messages (accepting or rejecting) a value during recovery.

Reply:
Note that the same safe guard discussed in https://github.com/cometbft/cometbft/issues/1035 should prevent nodes from equivocating in recovery mode.

Reply:
Yes, added a section on this